[GenericEnvironment] Include original parameter packs in opened pack element signatures.

hborla opened the pull request:
Model opened pack element signatures as a complete clone of the outer context signature, with an additional set of element type parameters at depth + 1 and corresponding element requirements. The generic environment stores a substitution map for all generic parameters whose depth is lower than the element depth, using the archetypes from the outer generic context when mapping types into this generic context.

For example, given this generic signature:

```swift
<T..., U> where T: P, T.A == U
```

The opened element generic signature is:

```swift
<T_0_0..., T_0_1, T_1_0> where T_0_0: P, T_1_0: P, T_0_0.[P]A == T_0_1, T_1_0.[P]A == T_0_1
```

where `T_1_0` represents an element in `T_0_0`. The element generic environment also stores substitutions for `T_0_0` and `T_0_1` to the archetypes created in the generic context of the original signature.

hborla committed:
[GenericEnvironment] Model opened pack element signatures as a complete clone
of the outer context signature, with an additional set of element type parameters
at depth + 1.

hborla committed:
[GenericEnvironment] Store a substitution map for the outer generic parameters
of an opened pack element generic environment.

hborla committed:
[GenericEnvironment] Add a method to retrieve the bindings from pack element
archetypes to their originating pack archetypes.

hborla committed:
[GenericEnvironment] Handle parameter packs in opened element signatures in
getOrCreateArchetypeFromInterfaceType.

Instead of special casing parameter packs for primary generic environments,
handle parameter packs in each kind of environment. For opened element
environments, a pack archetype can be produced from the outer substitutions.

hborla committed:
[RequirementMachine] Don't produced concrete type parameter diagnostics for
same-type requirements where one side is a parameter pack.

slavapestov commented on lib/AST/ASTVerifier.cpp:
```diff
@@ -634,10 +634,21 @@ class Verifier : public ASTWalker {
+            if (genericEnv->getKind() == GenericEnvironment::Kind::OpenedElement) {
+              auto contextSubs = genericEnv->getPackElementContextSubstitutions();
+              QuerySubstitutionMap isInContext{contextSubs};
+              if (isInContext(root->getInterfaceType()->getAs<GenericTypeParamType>()))
```
Should this be castTo<>?

hborla committed:
[ConstraintSystem] Correct the first type of the ShapeOf constraint when
generating constraints for PackExpansionExpr.

The result pattern type of a pack expansion doesn't always contain pack
references. Instead, use the type of the first pack reference binding.

hborla committed:
[ASTVerifier] Change 'getAs' to 'castTo' when verifying archetypes in
opened pack element environments.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test
